test/test-test.js

```diff
@@ -233,6 +233,69 @@
             }).call({}, "arg1", {}, done);
         },
 
+        "async promise-based test": {
+            requiresSupportFor: {
+                Promise: typeof Promise !== "undefined"
+            },
+
+            "restores the sandbox when the promise is fulfilled": function (done) {
+                var method = function () {};
+                var object = { method: method };
+                var spy = sinon.spy();
+                sinon.test(function (callback) {
+                    var stub;
+                    var sandbox = this;
+                    return Promise.resolve()
+                        .then(function () {
```

Comment:
> Remove spies from the equation. We're only testing the sandbox functionality.

+1

> Simply `return Promise.resolve();`.

Does that guarantee that the sandbox was restored after the Promise was resolved? I think it's important to keep the `sandbox.stub()` (or whatever sandbox method) in the `then` callback.